ДЗ по архитектуре

dimazollo opened the pull request:
В рамках домашнего задания реализован собственный Flux фреймворк. Описание API можно найти в README проекта.

Commit:
code my own flux framework and connect it to application

Commit:
Update README.md for the task

shri-msk-2018-reviewer commented on README.md:
```diff
@@ -12,9 +10,28 @@
+Для работы фреймворка необходимо создать объект класса MyFLux. 
+Конструктор этого класса принимает на вход три параметра:
+ - state: ILooseObject - объект, описывающий начальное состояние хранилища
+ - actions: Array<{type: string, handler: IActionHandler}> - массив, состоящий из объектов с парами свойств "тип действия", "обработчик действия".
```
Тайпскрипт - это хорошо!

shri-msk-2018-reviewer commented on src/js/app.ts:
```diff
@@ -1,32 +1,40 @@
+  return newElement;
+});
+
+const myFlux = new MyFlux(state, actions, [eventsComponent]);
```
Верно :-)

shri-msk-2018-reviewer commented on src/js/flux/Action.ts:
```diff
@@ -0,0 +1,21 @@
+import ILooseObject from './interface/iLooseObject';
+
+export default class Action {
+  private readonly _type: string = '';
```
верно :-)

shri-msk-2018-reviewer commented on src/js/flux/Dispatcher.ts:
```diff
@@ -0,0 +1,25 @@
+import Store from './Store';
+import Action from './Action';
+
+export default class Dispatcher {
```
вопрос на засыпку - как называется этот паттерн?

dimazollo replied:
Смею предположить, что это паттерн Mediator

shri-msk-2018-reviewer replied:
Думаю, нет, хотя можем поспорить. Это та самая инверсия зависимостей, которая была в лекции - один в один.

shri-msk-2018-reviewer commented on src/js/flux/Store.ts:
```diff
@@ -0,0 +1,28 @@
+
+  public apply(action: Action) {
+    this.state = this.handleAction(this.state, action);
+    document.dispatchEvent(new CustomEvent(fluxChangeEvent));
```
здесь можно было написать eventEmitter и не пользоваться механизмом document

shri-msk-2018-reviewer commented on src/js/flux/Store.ts:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  public setState(newState: ILooseObject) {
+    this.state = newState;
```
вот это нехорошо. Внешний метод, который может изменить state стора - это очень небезопасно. Нужно, чтобы стор сам дергал метод изменения

dimazollo replied:
Не помню почему я решил этот метод добавить, ведь сам я им в работе не воспользовался. Наверное добавил пока еще не разобрался.

shri-msk-2018-reviewer commented on src/js/flux/View.ts:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  setStore(store: Store) {
+    this.store = store;
```
Это все-таки state наверное? Store обычно один

shri-msk-2018-reviewer commented on src/js/store.ts:
```diff
@@ -0,0 +1,31 @@
+import IActionHandler from './flux/interface/IActionHandler';
```
А этот store уже для проекта? Чуть не запуталась в названиях

shri-msk-2018-reviewer commented on src/js/store.ts:
```diff
@@ -0,0 +1,31 @@
+    handler: {
+      handleAction(previousState, payload) {
+        const newState = Object.assign({}, previousState);
+        newState.events = newState.events.filter(
```
можно сделать events хэшом или Map и искать по id без фильтрации

Commit:
remove unsafe method setStore

Commit:
remove redundant code and comments

Commit:
add event emitter

Commit:
remove property DI of Store to every registered View and replace this type of
coupling using EventEmitter and subscriptions of View to broadcasting change event.
The event also provides each View a copy of actual state.

Commit:
fix problem with passing actual state to action handlers, not copy of it

shri-msk-2018-reviewer commented on src/js/flux/EventEmitter.ts:
```diff
@@ -0,0 +1,24 @@
+
+export default class EventEmitter {
+  private subscribers: {
+    [eventType: string]: Array<View>;
```
У нас подписываться могут только вьюшки? А если мы захотим использовать эмиттер в другом месте?

Подсказка: если не хочется писать `Array<any>` - воспользуйтесь дженериком

dimazollo replied:
Создал интерфейс IEventHandler, которым заменил View внутри EventEmitter, и который теперь имплементит View

shri-msk-2018-reviewer commented on src/js/flux/EventEmitter.ts:
```diff
@@ -0,0 +1,24 @@
+
+  unsubscribe(eventType: string, view: View) {
+    this.subscribers[eventType] =
+      this.subscribers[eventType].filter(subscriber => subscriber !== view);
```
а можно обойтись здесь без фильтра? Сразу найти переданный view в списке подписчиков?

shri-msk-2018-reviewer commented on src/js/flux/EventEmitter.ts:
```diff
@@ -0,0 +1,24 @@
+  broadcast(eventType:string, state: ILooseObject) {
+    this.subscribers[eventType].forEach(subscriber => subscriber.update(state));
+  }
+}
```
В целом реализация корректная, комментарии тут про красоту :-)

shri-msk-2018-reviewer commented on src/js/flux/Store.ts:
```diff
@@ -0,0 +1,28 @@
+
+  public apply(action: Action) {
+    this.state = this.handleAction(Object.assign({}, this.state), action);
+    this.eventEmitter.broadcast(
```
верно :-)

shri-msk-2018-reviewer commented on src/js/store.ts:
```diff
@@ -0,0 +1,29 @@
+    type: 'close-event',
+    handler: {
+      handleAction(storeState, payload) {
+        storeState.events = storeState.events.filter(
```
вот это у вас редьюсер (в терминологии redux). В принципе его можно вынести в отдельную сущность - так удобнее работать с фреймворком. Но концептуально стало правильно.

Commit:
Remove straight dependency on View from EventEmitter by replacing View to IEventHandler interface.
Implement IEventHandler interface in View
